### PR TITLE
LibraryApi done and up to par with FollowApi style

### PR DESCRIPTION
**Pull request by akshays2112**
LibraryApi is done and up to par with FollowApi style of coding and commenting.

### CODE REVIEW

File: src/SpotifyApi.NetCore/LibraryApi.cs [hunk 1]
```diff
@@ -0,0 +1,410 @@
...
+        /// </summary>
+        /// <param name="albumIds">Required. A comma-separated list of the Spotify IDs for the albums. Minimum: 1 ID. Maximum: 50 IDs.</param>
+        /// <returns>bool[] an array of true or false values, in the same order in which the ids were specified.</returns>
+        /// <remarks>
+        /// https://developer.spotify.com/documentation/web-api/reference/library/check-users-saved-albums/
+        /// </remarks>
+        public async Task<bool[]> CheckUserSavedAlbums(
+            string[] albumIds,
+            string accessToken = null
+            ) => await CheckUserSavedAlbums<bool[]>(albumIds, accessToken);
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/LibraryApi.cs]: * Method name should be `CheckUsersSavedAlbums`
* Don't `async` `await` here. No need to unwind the Task. This goes for all non-generic methods that are proxies for the Generic version 🙂

File: src/SpotifyApi.NetCore/LibraryApi.cs [hunk 2]
```diff
@@ -0,0 +1,410 @@
...
+        /// </summary>
+        /// <param name="showIds">Required. A comma-separated list of the Spotify IDs for the shows. Minimum: 1 ID. Maximum: 50 IDs.</param>
+        /// <returns>bool[] an array of true or false values, in the same order in which the ids were specified.</returns>
+        /// <remarks>
+        /// https://developer.spotify.com/documentation/web-api/reference/library/check-users-saved-shows/
+        /// </remarks>
+        public async Task<bool[]> CheckUserSavedShows(
+            string[] showIds,
+            string accessToken = null
+            ) => await CheckUserSavedShows<bool[]>(showIds, accessToken);
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/LibraryApi.cs]: * Method name should be `CheckUsersSavedShows`
* Remove `async` and `await` keywords here.

File: src/SpotifyApi.NetCore/LibraryApi.cs [hunk 3]
```diff
@@ -0,0 +1,410 @@
...
+        /// </summary>
+        /// <param name="trackIds">Required. A comma-separated list of the Spotify IDs for the tracks. Minimum: 1 ID. Maximum: 50 IDs.</param>
+        /// <returns>bool[] an array of true or false values, in the same order in which the ids were specified.</returns>
+        /// <remarks>
+        /// https://developer.spotify.com/documentation/web-api/reference/library/check-users-saved-tracks/
+        /// </remarks>
+        public async Task<bool[]> CheckUserSavedTracks(
+            string[] trackIds,
+            string accessToken = null
+            ) => await CheckUserSavedTracks<bool[]>(trackIds, accessToken);
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/LibraryApi.cs]: * Method name = `CheckUsersSavedTracks`
* Remove `async` and `await` keywords here.

File: src/SpotifyApi.NetCore/LibraryApi.cs [hunk 4]
```diff
@@ -0,0 +1,410 @@
...
+        /// <returns>A Task that, once successfully completed, returns a full <see cref="PagedAlbums"/> object.</returns>
+        /// <remarks>
+        /// https://developer.spotify.com/documentation/web-api/reference/library/get-users-saved-albums/
+        /// </remarks>
+        public async Task<PagedAlbums> GetUserSavedAlbums(
+            int limit = 20,
+            int offset = 0,
+            string market = null,
+            string accessToken = null
+            ) => await GetUserSavedAlbums<PagedAlbums>(limit, offset, market, accessToken);
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/LibraryApi.cs]: Remove `async` and `await` keywords here.

File: src/SpotifyApi.NetCore/LibraryApi.cs [hunk 5]
```diff
@@ -0,0 +1,410 @@
...
+
+        /// <summary>
+        /// Check if one or more albums is already saved in the current Spotify user’s ‘Your Music’ library.
+        /// </summary>
+        /// <param name="albumIds">Required. A comma-separated list of the Spotify IDs for the albums. Minimum: 1 ID. Maximum: 50 IDs.</param>
+        /// <returns>bool[] an array of true or false values, in the same order in which the ids were specified.</returns>
+        /// <remarks>
+        /// https://developer.spotify.com/documentation/web-api/reference/library/check-users-saved-albums/
+        /// </remarks>
+        public async Task<T> CheckUserSavedAlbums<T>(
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/LibraryApi.cs]: Method name should be `CheckUsersSavedAlbums`

File: src/SpotifyApi.NetCore/LibraryApi.cs [hunk 6]
```diff
@@ -0,0 +1,410 @@
...
+
+        /// <summary>
+        /// Check if one or more shows is already saved in the current Spotify user’s library.
+        /// </summary>
+        /// <param name="showIds">Required. A comma-separated list of the Spotify IDs for the shows. Minimum: 1 ID. Maximum: 50 IDs.</param>
+        /// <returns>bool[] an array of true or false values, in the same order in which the ids were specified.</returns>
+        /// <remarks>
+        /// https://developer.spotify.com/documentation/web-api/reference/library/check-users-saved-shows/
+        /// </remarks>
+        public async Task<T> CheckUserSavedShows<T>(
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/LibraryApi.cs]: Method name should be `CheckUsersSavedShows`

File: src/SpotifyApi.NetCore/LibraryApi.cs [hunk 7]
```diff
@@ -0,0 +1,410 @@
...
+
+        /// <summary>
+        /// Check if one or more tracks is already saved in the current Spotify user’s ‘Your Music’ library.
+        /// </summary>
+        /// <param name="trackIds">Required. A comma-separated list of the Spotify IDs for the tracks. Minimum: 1 ID. Maximum: 50 IDs.</param>
+        /// <returns>bool[] an array of true or false values, in the same order in which the ids were specified.</returns>
+        /// <remarks>
+        /// https://developer.spotify.com/documentation/web-api/reference/library/check-users-saved-tracks/
+        /// </remarks>
+        public async Task<T> CheckUserSavedTracks<T>(
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/LibraryApi.cs]: Method name = `CheckUsersSavedTracks`

File: src/SpotifyApi.NetCore/LibraryApi.cs [hunk 8]
```diff
@@ -0,0 +1,410 @@
...
+        /// Get a list of the albums saved in the current Spotify user’s ‘Your Music’ library.
+        /// </summary>
+        /// <param name="limit">Optional. The maximum number of objects to return. Default: 20. Minimum: 1. Maximum: 50.</param>
+        /// <param name="offset">Optional. The index of the first object to return. Default: 0 (i.e., the first object). Use with limit to get the next set of objects.</param>
+        /// <param name="market">Optional. An ISO 3166-1 alpha-2 country code or the string from_token. Provide this parameter if you want to apply Track Relinking.</param>
+        /// <returns>A Task that, once successfully completed, returns a full <see cref="PagedAlbums"/> object.</returns>
+        /// <remarks>
+        /// https://developer.spotify.com/documentation/web-api/reference/library/get-users-saved-albums/
+        /// </remarks>
+        public async Task<PagedAlbums> GetUserSavedAlbums(
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/LibraryApi.cs]: * Method name = `GetCurrentUsersSavedAlbums`
* Remove `async` `await` keywords

File: src/SpotifyApi.NetCore/LibraryApi.cs [hunk 9]
```diff
@@ -0,0 +1,410 @@
...
+        }
+        #endregion
+
+        #region GetUserSavedAlbums
+        /// <summary>
+        /// Get a list of the albums saved in the current Spotify user’s ‘Your Music’ library.
+        /// </summary>
+        /// <param name="limit">Optional. The maximum number of objects to return. Default: 20. Minimum: 1. Maximum: 50.</param>
+        /// <param name="offset">Optional. The index of the first object to return. Default: 0 (i.e., the first object). Use with limit to get the next set of objects.</param>
+        /// <param name="market">Optional. An ISO 3166-1 alpha-2 country code or the string from_token. Provide this parameter if you want to apply Track Relinking.</param>
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/LibraryApi.cs]: Mention the `SpotifyCountryCodes` helper here e.g.

```csharp
/// <param name="market">Optional. An ISO 3166-1 alpha-2 country code or the string from_token <see cref="SpotifyCountryCodes" />. Provide this parameter if you want to apply Track Relinking.</param>
```

File: src/SpotifyApi.NetCore/LibraryApi.cs [hunk 10]
```diff
@@ -0,0 +1,410 @@
...
+        /// <returns>A Task that, once successfully completed, returns a full <see cref="PagedAlbums"/> object.</returns>
+        /// <remarks>
+        /// https://developer.spotify.com/documentation/web-api/reference/library/get-users-saved-albums/
+        /// </remarks>
+        public async Task<T> GetUserSavedAlbums<T>(
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/LibraryApi.cs]: Method name = `GetCurrentUsersSavedAlbums`

File: src/SpotifyApi.NetCore/LibraryApi.cs [hunk 11]
```diff
@@ -0,0 +1,410 @@
...
+            string market = null,
+            string accessToken = null
+            ) => await GetUserSavedAlbums<PagedAlbums>(limit, offset, market, accessToken);
+
+        /// <summary>
+        /// Get a list of the albums saved in the current Spotify user’s ‘Your Music’ library.
+        /// </summary>
+        /// <param name="limit">Optional. The maximum number of objects to return. Default: 20. Minimum: 1. Maximum: 50.</param>
+        /// <param name="offset">Optional. The index of the first object to return. Default: 0 (i.e., the first object). Use with limit to get the next set of objects.</param>
+        /// <param name="market">Optional. An ISO 3166-1 alpha-2 country code or the string from_token. Provide this parameter if you want to apply Track Relinking.</param>
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/LibraryApi.cs]: Mention the `SpotifyCountryCodes` helper here e.g.

```csharp
/// <param name="market">Optional. An ISO 3166-1 alpha-2 country code or the string from_token <see cref="SpotifyCountryCodes" />. Provide this parameter if you want to apply Track Relinking.</param>
```

File: src/SpotifyApi.NetCore/LibraryApi.cs [hunk 12]
```diff
@@ -0,0 +1,410 @@
...
+        /// <summary>
+        /// Get a list of shows saved in the current Spotify user’s library. Optional parameters can be used to limit the number of shows returned.
+        /// </summary>
+        /// <param name="limit">Optional. The maximum number of objects to return. Default: 20. Minimum: 1. Maximum: 50.</param>
+        /// <param name="offset">Optional. The index of the first object to return. Default: 0 (i.e., the first object). Use with limit to get the next set of objects.</param>
+        /// <returns>A Task that, once successfully completed, returns a full <see cref="PagedShows"/> object.</returns>
+        /// <remarks>
+        /// https://developer.spotify.com/documentation/web-api/reference/library/get-users-saved-shows/
+        /// </remarks>
+        public async Task<PagedShows> GetUserSavedShows(
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/LibraryApi.cs]: * Method name = `GetUsersSavedShows`
* Remove `async` and `await` keywords here

File: src/SpotifyApi.NetCore/LibraryApi.cs [hunk 13]
```diff
@@ -0,0 +1,410 @@
...
+        /// <summary>
+        /// Get User's Saved Shows
+        /// </summary>
+        /// <param name="limit">Optional. The maximum number of objects to return. Default: 20. Minimum: 1. Maximum: 50.</param>
+        /// <param name="offset">Optional. The index of the first object to return. Default: 0 (i.e., the first object). Use with limit to get the next set of objects.</param>
+        /// <returns>A Task that, once successfully completed, returns a full <see cref="PagedShows"/> object.</returns>
+        /// <remarks>
+        /// https://developer.spotify.com/documentation/web-api/reference/library/get-users-saved-shows/
+        /// </remarks>
+        public async Task<T> GetUserSavedShows<T>(
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/LibraryApi.cs]: Method name = `GetUsersSavedShows`

File: src/SpotifyApi.NetCore/LibraryApi.cs [hunk 14]
```diff
@@ -0,0 +1,410 @@
...
+        /// Get a list of the songs saved in the current Spotify user’s ‘Your Music’ library.
+        /// </summary>
+        /// <param name="limit">Optional. The maximum number of objects to return. Default: 20. Minimum: 1. Maximum: 50.</param>
+        /// <param name="offset">Optional. The index of the first object to return. Default: 0 (i.e., the first object). Use with limit to get the next set of objects.</param>
+        /// <param name="market">Optional. An ISO 3166-1 alpha-2 country code or the string from_token. Provide this parameter if you want to apply Track Relinking.</param>
+        /// <returns>A Task that, once successfully completed, returns a full <see cref="PagedTracks"/> object.</returns>
+        /// <remarks>
+        /// https://developer.spotify.com/documentation/web-api/reference/library/get-users-saved-tracks/
+        /// </remarks>
+        public async Task<PagedTracks> GetUserSavedTracks(
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/LibraryApi.cs]: * Method name = `GetUsersSavedTracks`
* Remove `async` and `await` keywords here

File: src/SpotifyApi.NetCore/LibraryApi.cs [hunk 15]
```diff
@@ -0,0 +1,410 @@
...
+        }
+        #endregion
+
+        #region GetUserSavedTracks
+        /// <summary>
+        /// Get a list of the songs saved in the current Spotify user’s ‘Your Music’ library.
+        /// </summary>
+        /// <param name="limit">Optional. The maximum number of objects to return. Default: 20. Minimum: 1. Maximum: 50.</param>
+        /// <param name="offset">Optional. The index of the first object to return. Default: 0 (i.e., the first object). Use with limit to get the next set of objects.</param>
+        /// <param name="market">Optional. An ISO 3166-1 alpha-2 country code or the string from_token. Provide this parameter if you want to apply Track Relinking.</param>
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/LibraryApi.cs]: Mention the `SpotifyCountryCodes` helper here e.g.

```csharp
/// <param name="market">Optional. An ISO 3166-1 alpha-2 country code or the string from_token <see cref="SpotifyCountryCodes" />. Provide this parameter if you want to apply Track Relinking.</param>
```

File: src/SpotifyApi.NetCore/LibraryApi.cs [hunk 16]
```diff
@@ -0,0 +1,410 @@
...
+        /// <returns>A Task that, once successfully completed, returns a full <see cref="PagedTracks"/> object.</returns>
+        /// <remarks>
+        /// https://developer.spotify.com/documentation/web-api/reference/library/get-users-saved-tracks/
+        /// </remarks>
+        public async Task<T> GetUserSavedTracks<T>(
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/LibraryApi.cs]: * Method name = `GetUsersSavedTracks`

File: src/SpotifyApi.NetCore/LibraryApi.cs [hunk 17]
```diff
@@ -0,0 +1,410 @@
...
+            string market = null,
+            string accessToken = null
+            ) => await GetUserSavedTracks<PagedTracks>(limit, offset, market, accessToken);
+
+        /// <summary>
+        /// Get a list of the songs saved in the current Spotify user’s ‘Your Music’ library.
+        /// </summary>
+        /// <param name="limit">Optional. The maximum number of objects to return. Default: 20. Minimum: 1. Maximum: 50.</param>
+        /// <param name="offset">Optional. The index of the first object to return. Default: 0 (i.e., the first object). Use with limit to get the next set of objects.</param>
+        /// <param name="market">Optional. An ISO 3166-1 alpha-2 country code or the string from_token. Provide this parameter if you want to apply Track Relinking.</param>
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/LibraryApi.cs]: Mention the `SpotifyCountryCodes` helper here e.g.

```csharp
/// <param name="market">Optional. An ISO 3166-1 alpha-2 country code or the string from_token <see cref="SpotifyCountryCodes" />. Provide this parameter if you want to apply Track Relinking.</param>
```

File: src/SpotifyApi.NetCore/LibraryApi.cs [hunk 18]
```diff
@@ -0,0 +1,410 @@
...
+        #region RemoveUserSavedShows
+        /// <summary>
+        /// Delete one or more shows from current Spotify user’s library.
+        /// </summary>
+        /// <param name="showIds">Required. A comma-separated list of the show Spotify IDs. A maximum of 50 show IDs can be sent in one request. A minimum of 1 show id is required.</param>
+        /// <param name="market">Optional. An ISO 3166-1 alpha-2 country code. If a country code is specified, only shows that are available in that market will be removed. If a valid user access token is specified in the request header, the country associated with the user account will take priority over this parameter. Note: If neither market or user country are provided, the content is considered unavailable for the client. Users can view the country that is associated with their account in the account settings.</param>
+        /// <remarks>
+        /// https://developer.spotify.com/documentation/web-api/reference/library/remove-shows-user/
+        /// </remarks>
+        public async Task RemoveUserSavedShows(
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/LibraryApi.cs]: Method name = `RemoveUsersSavedShows`

File: src/SpotifyApi.NetCore/LibraryApi.cs [hunk 19]
```diff
@@ -0,0 +1,410 @@
...
+            await Delete(builder.Uri, accessToken);
+        }
+        #endregion
+
+        #region RemoveUserSavedShows
+        /// <summary>
+        /// Delete one or more shows from current Spotify user’s library.
+        /// </summary>
+        /// <param name="showIds">Required. A comma-separated list of the show Spotify IDs. A maximum of 50 show IDs can be sent in one request. A minimum of 1 show id is required.</param>
+        /// <param name="market">Optional. An ISO 3166-1 alpha-2 country code. If a country code is specified, only shows that are available in that market will be removed. If a valid user access token is specified in the request header, the country associated with the user account will take priority over this parameter. Note: If neither market or user country are provided, the content is considered unavailable for the client. Users can view the country that is associated with their account in the account settings.</param>
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/LibraryApi.cs]: Mention the `SpotifyCountryCodes` helper here e.g.

```csharp
/// <param name="market">Optional. An ISO 3166-1 alpha-2 country code or the string from_token <see cref="SpotifyCountryCodes" />. Provide this parameter if you want to apply Track Relinking.</param>
```

File: src/SpotifyApi.NetCore/LibraryApi.cs [hunk 20]
```diff
@@ -0,0 +1,410 @@
...
+
+        #region RemoveUserSavedTracks
+        /// <summary>
+        /// Remove one or more tracks from the current user’s ‘Your Music’ library.
+        /// </summary>
+        /// <param name="trackIds">Required. A comma-separated list of the track Spotify IDs. A maximum of 50 track IDs can be sent in one request. A minimum of 1 track id is required.</param>
+        /// <remarks>
+        /// https://developer.spotify.com/documentation/web-api/reference/library/remove-tracks-user/
+        /// </remarks>
+        public async Task RemoveUserSavedTracks(
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/LibraryApi.cs]: Method name = `RemoveUsersSavedTracks`

File: src/SpotifyApi.NetCore/LibraryApi.cs [hunk 21]
```diff
@@ -0,0 +1,410 @@
...
+
+        #region SaveTracksForCurrentUser
+        /// <summary>
+        /// Save one or more tracks to the current user’s ‘Your Music’ library.
+        /// </summary>
+        /// <param name="trackIds">Required. A comma-separated list of the track Spotify IDs. A maximum of 50 track IDs can be sent in one request. A minimum of 1 track id is required.</param>
+        /// <remarks>
+        /// https://developer.spotify.com/documentation/web-api/reference/library/save-tracks-user/
+        /// </remarks>
+        public async Task SaveTracksForCurrentUser(
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/LibraryApi.cs]: Method name = `SaveTracksForUser`

File: src/SpotifyApi.NetCore/ILibraryApi.cs [hunk 22]
```diff
@@ -0,0 +1,285 @@
+﻿using SpotifyApi.NetCore.Models;
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/ILibraryApi.cs:1]: See review comments on `LibraryApi.cs` for changes to this interface

File: src/SpotifyApi.NetCore.Tests/LibraryApiTests.cs [hunk 23]
```diff
@@ -0,0 +1,272 @@
...
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using SpotifyApi.NetCore.Authorization;
+using SpotifyApi.NetCore.Models;
+
+namespace SpotifyApi.NetCore.Tests
+{
+    [TestClass]
+    public class LibraryApiTests
+    {
+        [TestCategory("Integration")]
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore.Tests/LibraryApiTests.cs:14]: Please add User test category to all tests as they require User auth, e.g.

```csharp
[TestCategory("Integration")]
[TestCategory("User")]
```

**Review by DanielLarsenNZ — CHANGES_REQUESTED** [review]
This is awesome, thank you @akshays2112. Just a few naming changes to match Spotify docs (which are inconsistent I know, but I want to follow them). Also some unnecessary use of `async` `await` keywords - this is minor but we should stay consistent with other classes.

**Comment by akshays2112** [comment]
Hi Daniel,

I got it. I am on it right away as it will not take much time.  I am watching this brilliant interviewer and of course Ronald Regan who is giving very smart responses to his questions.

Thanks for your patience and time.

Best Regards,

Akshay Srinivasan

**Comment by akshays2112** [comment]
Hi Daniel,

Sorry took so long. I am beggining now via pattern recognition to understand why these naming conventions of mine are going wrong.  Also the problem that came out of my understanding is the Spotify API naming conventions are inconsistent.  If you are using CurrentUser for Albumns but only using User for Shows then I expect that there should be a different meaning when one time they say Current User and when they say User only. This is why I was getting it wrong.

Thanks for your patience and time.

Best Regards,

Akshay Srinivasan

**Review by DanielLarsenNZ — APPROVED** [review]
BOOM 💥

**Comment by akshays2112** [comment]
♥♥♥♥

**Comment by akshays2112** [comment]
Hi Daniel,

I could not figure out your logic of which test to mark as [TestCategory("User")].
First I thought by looking at your modifications to FollowApiTests that it is maybe only if the current user account is being modified i.e. the current user account is the target of the action.
Or maybe is it when the user account is being accessed either read or write in any form.
Please clarify this soon as you see this. I am awaiting your response and then can finish off everything and create pull request for EpisodesApi.

Thanks for your patience and time.

Best Regards,

Akshay Srinivasan

**Pull request merged**